lib/native/http.ts

```diff
@@ -18,31 +18,69 @@ import { NativeResource } from "./native_resource";
 import { ResourceSafe } from '../common/resource_safety';
 import { ClientBootstrap, ClientTlsContext, SocketOptions } from './io';
 
-type ConnectionCallback = (error_code: Number) => void;
-
 export class HttpConnection extends NativeResource implements ResourceSafe {
 
-    constructor(
+    protected constructor(native_handle: any)
+    {
+        super(native_handle);
+    }
+
+    close() {
+        crt_native.http_connection_close(this.native_handle());
+    }
+}
+
+type ClientConnectionCallback = (connection: HttpClientConnection, error_code: Number) => void;
+
+export class HttpClientConnection extends HttpConnection {
+    static create(
         bootstrap: ClientBootstrap,
-        on_connection_setup: ConnectionCallback | undefined,
-        on_connection_shutdown: ConnectionCallback | undefined,
+        on_connection_setup: ClientConnectionCallback | undefined,
+        on_connection_shutdown: ClientConnectionCallback | undefined,
         host_name: String,
         port: Number,
         socket_options: SocketOptions,
-        tls_ctx: ClientTlsContext)
-    {
-        super(crt_native.http_connection_new(
-            bootstrap.native_handle(),
-            on_connection_setup,
-            on_connection_shutdown,
-            host_name,
-            port,
-            socket_options.native_handle(),
-            tls_ctx.native_handle()
-        ));
+        tls_ctx: ClientTlsContext) {
+        
+        return new Promise<HttpConnection>((resolve) => {
+            let connection: HttpClientConnection;
+            const on_setup = (native_connection: any, error_code: Number) => {
+                connection = new HttpClientConnection(
```

Comment:
whoa, you can capture a variable that hasn't been set yet.

black

magic

Reply:
That one took me a minute to come up with. Futures create a time paradox.